mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +347,206 @@ propagateUnmergeVectorization(T &&dimAndLength,
   return result;
 }
 
+// Key data structures for vectorization analysis
+using DimToMergeMap =
+    llvm::DenseMap<uint32_t,
+                   std::tuple<TransformMapAttr, TransformAttr, size_t>>;
+
+using ContiguousMergesMap =
+    llvm::DenseMap<std::pair<TransformMapAttr, TransformAttr>,
+                   SmallVector<SmallVector<uint32_t>>>;
+
+void findCountiguousGroupsUnmerge(const ArrayRef<uint32_t> upperDims,
+                                  const ArrayRef<int64_t> params,
+                                  DimToMergeMap &dimToMerge,
+                                  ContiguousMergesMap &contiguousGroups) {
+  size_t i = 0;
+  while (i < upperDims.size()) {
+    // The i-th dimension (i.e., dimension upperDims[i])
+    // uses params[i]
+
+    auto keyI = dimToMerge[upperDims[i]];
+    auto mergePairI = std::make_pair(std::get<0>(keyI), std::get<1>(keyI));
+    if (!mergePairI.second) {
+      i++;
+      continue;
+    }
+
+    SmallVector<uint32_t> groupCandidate;
+    SmallVector<size_t> dimPosition;
+    for (size_t j = i; j < upperDims.size(); j++) {
+
+      // Unit lengths don't affect the mergeability logic
+      if (params[j] == 1) {
```

Comment:
I'd actually argue for always adding unit dimensions to the current group. That way, every input to the Unmerge{} is inside a group

Reply:
But if you're only returning non-singleton groups, it works this way too.

The reason I suggested storing `contiguousGroups` as a `EquivalenceClasses` is because it makes the query logic inside vectorization easier.